### PR TITLE
Add support for multiple disks to testcloud plugin

### PR DESCRIPTION
**Pull request by happz**
```
TMT_SHOW_TRACEBACK=1 tmt -vv run plans --default \
    provision --how virtual --image fedora-39 \
              --hardware 'disk[1].size=20GB' \
              --hardware 'disk[0].size=15GB' \
    login -s provision \
    finish
```

Fixes https://github.com/teemtee/tmt/issues/2765

Pull Request Checklist

* [x] implement the feature
* [x] extend the test coverage
* [x] mention the version
* [x] include a release note

### CODE REVIEW

**Comment by happz** [comment]
@thrix right, I’ll add some test coverage, and I’ll use your examples too.

**Comment by thrix** [comment]
(shortened a bit for brewity)

> > /default/plan
> >     provision
> >         queued provision.provision task #1: default-0
> >         
> >         provision.provision task #1: default-0
> >         how: virtual
> >         hardware:
> >             disk:
> >               - {}
> >               - size: = 20GB
> >               - {}
> >               - {}
> >               - {}
> >               - {}
> >               - {}
> >               - {}
> >               - {}
> >               - {}
> >               - size: = 15GB
> 
> I'm open to suggestions for improvements.

I kinda understand why it looks like it, but from user perspective I am wondering if it is important that one disk is with index 1 and second with 10, looking at the booted machine, seems it is the same as if I would use index 0 and index 1?

Because of that I would go with:

> > /default/plan
> >     provision
> >         queued provision.provision task #1: default-0
> >         
> >         provision.provision task #1: default-0
> >         how: virtual
> >         hardware:
> >             disk:
> >               - size: = 20GB
> >               - size: = 15GB

**Comment by happz** [comment]
> @happz
> 
> (shortened a bit for brewity)
> 
> > > /default/plan
> > > provision
> > > queued provision.provision task #1: default-0
> > > ```
> > >     provision.provision task #1: default-0
> > >     how: virtual
> > >     hardware:
> > >         disk:
> > >           - {}
> > >           - size: = 20GB
> > >           - {}
> > >           - {}
> > >           - {}
> > >           - {}
> > >           - {}
> > >           - {}
> > >           - {}
> > >           - {}
> > >           - size: = 15GB
> > > ```
> > 
> > 
> > I'm open to suggestions for improvements.
> 
> I kinda understand why it looks like it, but from user perspective I am wondering if it is important that one disk is with index 1 and second with 10, looking at the booted machine, seems it is the same as if I would use index 0 and index 1?

Well, then you should have used `--hardware 'disk[0].size=15GB'` and `--hardware 'disk[1].size=15GB'`, not `--hardware 'disk[10].size=15GB'` ;)

Yeah, this is not well thought out, the order is much less important than it might look like. This might be actually a good idea, merge all inputs and then remove all empty slots. We need the index when defining the requirements, so you can refer to the same disk from multiple options (`--hardware 'disk[10].size=15GB' --hardware 'disk[10].vendor-name=Seagate'`), but after that, it's pointless.

> Because of that I would go with:
> 
> > > /default/plan
> > > provision
> > > queued provision.provision task #1: default-0
> > > ```
> > >     provision.provision task #1: default-0
> > >     how: virtual
> > >     hardware:
> > >         disk:
> > >           - size: = 20GB
> > >           - size: = 15GB
> > > ```

I'd like to achieve this:

```yaml
     hardware:
         disk[0].size: = 20GB
         disk[1].size: = 15GB
```

but it turns out to be hard to achieve with simple tools I have, it'd require way more work, so I will probably go with removing the empty slots.

**Comment by happz** [comment]
@thrix hopefully all three points were addressed in 2cf44e5f

**Review by therazix — APPROVED** [review]

**Review by KwisatzHaderach — APPROVED** [review]

**Review by martinhoyer — APPROVED** [review]
Nice, feel free to close the nitpicking comments.

File: tmt/hardware.py [hunk 1]
```diff
@@ -154,6 +154,21 @@ class Operator(enum.Enum):
...
     """, re.VERBOSE)
 
+
+#: A list of constraint names that operate over sequence of entities.
+INDEXABLE_CONSTRAINTS: tuple[str, ...] = (
+    'disk',
+    'network'
+    )
+
+#: A list oc constraint names that do not have child properties.
```
[the remark this answers is not in the document]
Comment (happz) [reply tmt/hardware.py:164]: Yeah, I tend to use "list" even for tuples, when they serve as collections of items. "Sequence" sounds too technical, and if it's a collection of items we iterate over, I often use "list" in documentation. I'd use "a tuple" when the order would matter, or when there were items of different types, like parameters for a unit test.

**Review by thrix — APPROVED** [review]
LGTM

**Comment by happz** [comment]
@thrix I'm putting placeholders back into output, I can't fix https://github.com/teemtee/tmt/issues/3004 without them for now. Good news is, you won't see them in real life that much, `disk[10]` is a very artificial example, therefore the output won't be as bad as it was here.

Added an issue to track it, and will follow up, not just to get rid of placeholders but also to clear up the position of "root disk" (or "boot disk"?).

**Comment by lukaszachy** [comment]
@happz Test failures are not related to this PR, right?

**Comment by martinhoyer** [comment]
> @happz Test failures are not related to this PR, right?

I don't think so fwiw.

**Comment by happz** [comment]
Unrelated, but the results for the `provision` were affected by something, and the test I updated got no results at all, which is weird and suspicious. Restarting rather than merging the PR.

**Pull request merged**